Усовершенствовал трекер задач.

ViktorSalk opened the pull request.

Viktor3642 committed:
В InMemoryHistoryManager:
   - Реализован двусвязный список с использованием класса Node.
   - Добавлен метод remove(int id) для удаления задачи из истории.
   - Добавлен метод removeNode(Node node) для удаления узла из списка.
   - Использована HashMap taskNodes для хранения соответствия id задачи и узла в списке.
   - Метод add(Task task) теперь удаляет дубликаты задач из истории.
В InMemoryTaskManager:
   - Добавлен вызов historyManager.remove(id) в методы deleteTask, deleteEpic и deleteSubtask для удаления задач из истории.
В Main:
   - Добавлен пользовательский сценарий для демонстрации работы истории.
   - (Дополнительное задание) Добавлен пользовательский сценарий, который демонстрирует работу истории, включая добавление задач, просмотр, удаление и вывод истории просмотров.
В InMemoryTaskManagerTest:
   - Добавлены тесты для проверки работы remove(int id) в InMemoryHistoryManager.
   - Добавлены тесты для проверки корректности удаления задач из истории в InMemoryTaskManager.
   - Обновлены существующие тесты для проверки корректности работы с историей.
   - Добавлены тесты для проверки целостности данных (отсутствие старых id в подзадачах и эпиках).
   - Добавлены тесты для проверки неизменяемости задач после их добавления в менеджер.

afomkina commented on src/main/java/manager/history/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,79 @@
+    private static class Node {
+        Task task;
+        Node next;
+        Node prev;
```
У полей класса должен быть модификатор доступа

ViktorSalk replied:
Спасибо. Добавил модификаторы доступа в поля  `Node next и Node prev.

afomkina commented on src/main/java/manager/history/InMemoryHistoryManager.java:
```diff
@@ -2,25 +2,79 @@
+    }
+
+    private static class Node {
+        Task task;
```
Поле должно быть final

ViktorSalk replied:
В Классе Node поле Task; теперь объявлено как "final" для обеспечения неизменности.

afomkina commented on src/main/java/Main.java:
```diff
@@ -91,5 +91,30 @@ public static void main(String[] args) {
+
+        // Пользовательский сценарий
+        System.out.println("\nПользовательский сценарий:");
+        taskManager.getTaskById(task1.getId());
```
Класс `TaskManager` нужно удалить из проекта и использовать "фабрику" `Managers`  метод `getDefault()` для создания экземпряра класса `InMemoryHistoryManager`

ViktorSalk replied:
Удалил класс `TaskManager` экземпляр теперь создается с использованием метода `Managers.getDefault()`.

afomkina commented on test/java/manager/history/task/manager/InMemoryTaskManagerTest.java:
```diff
@@ -206,4 +206,72 @@ void testHistoryManager() {
+
+    @Test
+    void testHistoryManagerWithDuplicates() {
+        TaskManager taskManager = Managers.getDefault();
```
Повторяющееся создание менеджера нужно вынести в переменные класса или метод [@BeforeEach](https://www.geeksforgeeks.org/junit-5-beforeeach/)

ViktorSalk replied:
Введен метод настройки "@beforeEach" в классе "InMemoryTaskManagerTest" и "HistoryTest".

afomkina commented on test/java/manager/history/task/manager/InMemoryTaskManagerTest.java:
```diff
@@ -206,4 +206,72 @@ void testHistoryManager() {
     }
+
+    @Test
+    void testHistoryManagerWithDuplicates() {
```
Тесты для проверкии испотрии нужно вынести в отдельный класс, чтобы не раздувать тесты по проверки логики менеджера задач

ViktorSalk replied:
Вывел в отдельный класс "HistoryTest" тесты для проверки истории.

afomkina reviewed (changes requested):
Нужно поправить структуру пакетов она должна быть следующей

```
src->
    main ->
        java ->
            manager     
            task
            history
    test ->
        java -> 
            manager
            task
            history
```

afomkina commented on src/main/java/manager/history/task/manager/InMemoryTaskManager.java:
```diff
@@ -72,6 +73,7 @@ public void deleteEpic(int id) {
         }
+        historyManager.remove(id);
     }
 
```
Метод `historyManager.remove(id)`; также должен использоваться в методах очищения задач по типам

ViktorSalk replied:
Добавлены вызовы "historyManager.remove(id)" в методах "clearTasks", "clearEpics" и "clearSubtasks" в классе "InMemoryTaskManager".

Viktor3642 committed:
1) Удалил класс `TaskManager` экземпляр теперь создается с использованием метода `Managers.getDefault()`
2) В Класс Node задача Task; теперь объявлено как "final" для обеспечения неизменности. Модификаторы доступа были добавлены в поля `Node next и Node prev для лучшей инкапсуляции.
3) Добавлены вызовы "historyManager.remove(id)" в методах "clearTasks", "clearEpics" и "clearSubtasks" в классе "InMemoryTaskManager".
4) Создан новый класс "HistoryTest", содержащий тесты, связанные с управлением историей, чтобы не загромождать тесты диспетчера задач.
5) Введен метод настройки "@beforeEach" в классе "InMemoryTaskManagerTest" и "HistoryTest", чтобы избежать повторного создания экземпляров `TaskManager` и `historyManager`.

Viktor3642 committed:
Поправил структуру пакетов, она стала более читаемой

afomkina commented:
UPD по структуре пакетов. Заметила, что тесты завязаны на местоположении папки test. Чтобы починить, структура должна быть следующей:
```
src ->
    history
    manager
    task
test ->
    history
    manager
    task
```

Viktor3642 committed:
Поправил структуру пакетов, для проходжения тестов.

afomkina commented on src/main/java/history/InMemoryHistoryManager.java:
```diff
@@ -0,0 +1,80 @@
+    private static class Node {
+        final Task task; // Заменено на final
+        Node next; // Добавлен модификатор доступа
+        Node prev; // Добавлен модификатор доступа
```
Нужно добавить модификатор доступа `private` для полей с 70 по 72 строку

ViktorSalk replied:
Спасибо) Добавил модификаторы доступа private для Task tsak, Nod next, Node prev.

afomkina commented on src/main/java/history/InMemoryHistoryManager.java:
```diff
@@ -0,0 +1,80 @@
+            } else {
+                tail.next = newNode;
+                newNode.prev = tail;
+                tail = newNode;
```
tail = newNode; является общей частью для обоих условий, нужно ее вынести за пределы проверки после 32 строки

ViktorSalk replied:
Убрал дубли выведя tail = newNode; за пределы проверки.

afomkina commented on test/java/manager/InMemoryTaskManagerTest.java:
```diff
@@ -1,21 +1,31 @@
-    private final TaskManager taskManager = Managers.getDefault();
-    private final HistoryManager historyManager = Managers.getDefaultHistory();
+    private TaskManager taskManager;
+    private HistoryManager historyManager;
```
Этот объект не используется в тесте, его нужно удалить

ViktorSalk replied:
Удалил неиспользуемые переменные в In Memory TaskManagerTest, по аналогии убрал неиспользуемые импорты в EpicTest, HistoryTest, InMemoryTaskManagerTest.

afomkina commented:
Нечаянно в трекере поставила не ту оценку, с проектом все хорошо, он работает и выглядит здорово, нужно немного поправить

Viktor3642 committed:
1) Добавил модификаторы доступа private для Task tsak, Nod next, Node prev.
2) Удалил неиспользуемые переменные в In Memory TaskManagerTest, и неиспользуемые импорты в EpicTest, HistoryTest, InMemoryTaskManagerTest.
3) Преобразовал условия выведя tail = newNode; за пределы проверки
